go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -238,14 +240,63 @@ func (w *heartbeatWriter) enableWrites(enable bool) {
 			w.ticks.Start(w.writeHeartbeat)
 		}()
 	case false:
-		w.ticks.Stop()
+		// We stop the ticks in a separate go routine because it can block if the write is stuck on semi-sync ACKs.
+		// At the same time we try and kill the write that is in progress. We use the context and its cancellation
+		// for coordination between the two go-routines. In the end we will have guaranteed that the ticks have stopped
+		// and no write is in progress.
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			w.ticks.Stop()
+			cancel()
+		}()
+		w.killWritesUntilStopped(ctx)
+
 		if w.onDemandDuration > 0 {
 			// Let the next RequestHeartbeats() go through
 			w.allowNextHeartbeatRequest()
 		}
 	}
 }
 
+// killWritesUntilStopped tries to kill the write in progress until the ticks have stopped.
+func (w *heartbeatWriter) killWritesUntilStopped(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			// If the context has been cancelled, then we know that the ticks have stopped.
+			// This guarantees that there are no writes in progress, so there is nothing to kill.
+			return
+		default:
+		}
+		// Actually try to kill the query.
+		err := w.killWrite()
+		w.recordError(err)
+		// We sleep for 100 millisecond, to prevent killing the query over and over very quickly.
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// killWrite kills the write in progress (if any).
+func (w *heartbeatWriter) killWrite() error {
+	defer w.env.LogError()
+	writeId := w.writeConnID.Load()
+	if writeId == -1 {
+		return nil
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), w.now().Add(w.interval))
+	defer cancel()
+	killConn, err := w.allPrivsPool.Get(ctx)
+	if err != nil {
+		log.Errorf("Kill conn didn't get connection :(")
+		return err
+	}
+	defer killConn.Recycle()
+
+	_, err = killConn.Conn.ExecuteFetch(fmt.Sprintf("kill %d", writeId), 1, false)
```

Comment:
I'm guessing `KILL QUERY %d` could also suffice here, but perhaps safer to just `KILL` the connection.

Reply:
Yes, I felt safer to just kill the entire connection.